cuappdev/old-eatery-blue-backend
# Finish implementing all schedule functionality (modified and closed schedules)

## Overview
This PR focuses on adding in functionality that enables eateries to modify their repeating schedules in the case of holidays or other exceptions. Also, it adds Alanna to CODEOWNERS so she also gets auto requested for PR review.

## Changes Made
- Created new model to deal with exceptions to repeated schedules, where each exception represents a modified or closed schedule
- Implemented a few new dfg nodes (ClosedSchedule, ModifiedSchedules, LeftMergeRepeatedEvents, and LeftMergeRegularEvents) to handle the manipulation of DB data into eatery menus for our API
- Refactored LeftMergeEvents to generalize it
- NOTE: I know the files I created for the nodes aren't named pythonically so let me know if I should re-name them before merging. I wanted to keep them as the same format as everything else to avoid confusion. Was planning on just refactoring all file names when we begin to restructure/decouple our codebase into MVC format but can also rename for this.


## Test Coverage
How I tested/how reviewers can test:
- Run bash reset_db.sh to handle restructuring the db file on your system to include new models
- Add new exceptions to api_scheduleexception table (need to do this manually since we don't have endpoints yet)
- Here's an example:
<img width="426" alt="Screen Shot 2022-04-17 at 5 56 15 PM" src="https://user-images.githubusercontent.com/22649594/163733471-bdce2c39-60c4-4578-8d61-7f068fdc4dbb.png">

- Ensure that the created exceptions go into effect -- if there is a repeating schedule it should be modified by the exception as intended
- Here is are examples of repeating schedules (also needs to be created in api_repeatingeventschedule table):
<img width="594" alt="Screen Shot 2022-04-17 at 5 57 19 PM" src="https://user-images.githubusercontent.com/22649594/163733528-e6b34fef-019b-42aa-bdf1-ac8ec45c390c.png">

## Related PRs or Issues (delete if not applicable)
PR to implement just repeating schedules: #20 

## Screenshots (delete if not applicable)

<details>

  <summary>The before v after of adding a closed schedule -- the parent event it points to, the one generated via repeating schedule 2, is removed (for testing modified schedules just make sure the start and end timestamps change as expected)</summary>

  
<img width="1423" alt="Screen Shot 2022-04-17 at 6 00 02 PM" src="https://user-images.githubusercontent.com/22649594/163733643-393e776a-dc4d-4b79-8449-6ea55d6e89f9.png">
<img width="1432" alt="Screen Shot 2022-04-17 at 6 00 08 PM" src="https://user-images.githubusercontent.com/22649594/163733647-d1916cb6-1e9c-43cf-9223-a2f80d14260a.png">


</details>
